enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,193 @@
+
+If the cleanup API is only present on the Subscription then operators installed without Subscriptions (by manually creating the CSV and fulfilling requirements) can not opt-in to operand cleanup.
+
+The CSV is the more appropriate point to configure the cleanup option for an operator as the CSV represents the actual installation of the operator, whereas the Subscription only represents the intent to subscribe to updates from the catalog. Moreover, since deleting the CSV causes the operator to be removed, deleting the CSV should be the trigger to initiate operand cleanup. See the [section on CSV finalizers][finalizers-on-csvs] for more details.
+
+#### Deleting CRs
+
+The operand cleanup process, simply put, is the removal of all CRs being reconciled by an operator so it can run finalizers for each CR and clean up any associated operator-created resources.
+
+From OLM's perspective these would be all CRs in the `OperatorGroup`'s [target namespaces][target-namespaces] that are of the same type as the [owned APIservices][owned-apiservices] listed in the CSV. This represents the set of all CRs that the operator reconciles and is responsible for cleaning up after.
+
+OLM should not delete CRs of the type [required APIService][required-apiservices] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - olm/operand-cleanup
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+With `spec.cleanup.enabled: false` OLM will not perform cleanup or delete any CRs. The finalizer on the CSV will be cleared so the CSV and operator can be removed without any cleanup. This is also the default behavior if `spec.cleanup` is unspecified to make cleanup opt-in.
```

Comment:
> The finalizer on the CSV will be cleared

...or never added in the first place?

Reply:
That's what I had in mind initially but with no finalizer in the default case, there could be a race condition when a user sets `cleanup.enabled: true` and then deletes the CSV before OLM has a chance to reconcile the CSV and set the finalizer to block deletion.

Although after considering this issue a bit more, I think I'd prefer this (seemingly unlikely) race condition as opposed to making OLM set finalizers for CSVs by default, and potentially affecting CSVs/operators that haven't opted into the cleanup behavior.

I'll update the proposal to make it so that the finalizer is only set after `cleanup.enabled: true`. And for `cleanup.enabled: false` or not specified the finalizer should be removed.